sql/src/main/java/io/crate/action/sql/TransportSQLAction.java

```diff
@@ -49,141 +42,189 @@
+        this.sqlOperations = sqlOperations;
         transportService.registerRequestHandler(SQLAction.NAME, SQLRequest.class, ThreadPool.Names.SAME, new TransportHandler());
     }
 
-    @Override
-    public ParameterContext getParamContext(SQLRequest request) {
-        return new ParameterContext(
-                request.args(), SQLBulkRequest.EMPTY_BULK_ARGS, request.getDefaultSchema(), request.getRequestFlags());
+    static Set<SQLOperations.Option> toOptions(int requestFlags) {
+        switch (requestFlags) {
+            case SQLBaseRequest.HEADER_FLAG_OFF:
+                return SQLOperations.Option.NONE;
+            case SQLBaseRequest.HEADER_FLAG_ALLOW_QUOTED_SUBSCRIPT:
+                return EnumSet.of(SQLOperations.Option.ALLOW_QUOTED_SUBSCRIPT);
+        }
+        throw new IllegalArgumentException("Unrecognized requestFlags: " + requestFlags);
     }
 
     @Override
-    void executePlan(Executor executor,
-                     final Analysis analysis,
-                     Plan plan,
-                     final ActionListener<SQLResponse> listener,
-                     final SQLRequest request,
-                     final long startTime) {
-
-        executor.execute(plan, new RowReceiver() {
-
-            final List<Object[]> rows = new ArrayList<>();
-            RowReceiver.Result setNextRowResult = RowReceiver.Result.CONTINUE;
-
-            @Override
-            public RowReceiver.Result setNextRow(Row row) {
-                rows.add(row.materialize());
-                return setNextRowResult;
+    protected void doExecute(SQLRequest request, ActionListener<SQLResponse> listener) {
+        SQLOperations.Session session = sqlOperations.createSession(
+            request.getDefaultSchema(), toOptions(request.getRequestFlags()), DEFAULT_SOFT_LIMIT);
+        try {
+            long startTime = System.nanoTime();
+            session.parse(UNNAMED, request.stmt(), Collections.<DataType>emptyList());
+            List<Object> args = request.args() == null ? Collections.emptyList() : Arrays.asList(request.args());
+            session.bind(UNNAMED, UNNAMED, args, null);
+            List<Field> outputFields = session.describe('P', UNNAMED);
+            if (outputFields == null) {
+                ResultReceiver resultReceiver = new RowCountReceiver(listener, startTime, request.includeTypesOnResponse());
+                session.execute(UNNAMED, 1, resultReceiver);
+            } else {
+                ResultReceiver resultReceiver = new ResultSetReceiver(listener, outputFields, startTime, request.includeTypesOnResponse());
+                session.execute(UNNAMED, 0, resultReceiver);
             }
+            session.sync(CompletionListener.NO_OP);
+        } catch (Throwable t) {
+            listener.onFailure(Exceptions.createSQLActionException(t));
+        }
+    }
 
-            @Override
-            public void pauseProcessed(ResumeHandle resumeable) {
+    public void disable() {
```

Comment:
The DecommissionService calls disable, but it never calls enable if it aborts - so that is probably bugged on master.
I think enable/disable should become part of the SQLOperations class but I thought I'd do that in a separate step/PR - after it is fixed.